src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -1337,4 +1357,156 @@ internal bool CanAddValue(HttpHeaderParser parser)
         }
         #endregion
     }
+
+    public readonly struct HttpHeadersNonValidated : IEnumerable<KeyValuePair<string, HeaderStringValues>>
+    {
+        private readonly IEnumerable<KeyValuePair<HeaderDescriptor, object>>? _headerStore;
+
+        internal HttpHeadersNonValidated(IEnumerable<KeyValuePair<HeaderDescriptor, object>>? headerStore)
+        {
+            _headerStore = headerStore;
+        }
+
+        public IEnumerator<KeyValuePair<string, HeaderStringValues>> GetEnumerator() => _headerStore != null
+            ? new Enumerator(_headerStore)
+            : ((IEnumerable<KeyValuePair<string, HeaderStringValues>>)Array.Empty<KeyValuePair<string, HeaderStringValues>>()).GetEnumerator();
+
+        IEnumerator<KeyValuePair<string, HeaderStringValues>> IEnumerable<KeyValuePair<string, HeaderStringValues>>.GetEnumerator() => GetEnumerator();
```

Comment:
Why do we need this? There's already public `GetEnumerator`. Doesn't it match the `IEnumerable<KeyValuePair<string, HeaderStringValues>>` contract?

Reply:
API proposal say that it should be returning `Enumerator`, but that's probably not possible.
If we're not exposing the `Enumerator`, couldn't it be private then?

Reply:
It's an explicit interface implementation. But the above `GetEnumerator()` can return `Enumerator`. Will try to fix it.

Reply:
Ah. No, it cannot, because it does a fast shortcut for an empty `_headerStore`.

Reply:
We want the public GetEnumerator to return the Enumerator struct so that it doesn't get boxed.

Reply:
This still needs to get fixed to return Enumerator here so we avoid boxing.

Reply:
The above GetEnumerator definition is fixed. This line cannot be changed because it's part of interface.
```C#
public Enumerator GetEnumerator() => new Enumerator(_headerStore);
```